services/management/commands/turku_service_import/addresses.py

```diff
@@ -0,0 +1,95 @@
+
+    def _import_address(self, entry):
+        street, _ = Street.objects.get_or_create(**entry['street'])
+        location = Point(
+            srid=3877,
+            **entry['point']
+        )
+
+        address, created = Address.objects.get_or_create(
+            street=street,
+            defaults={'location': location},
+            **entry['address']
+        )
+
+        if created:
+            address.refresh_from_db()
+
+    def _create_address_mapping(self, address_reader):
+        turku = Municipality.objects.get(id='turku')
+        multi_lingual_addresses = {}
+        for row in address_reader:
+            if row['city'].lower() not in self.valid_cities:  # Turku only
+                continue
+
+            coordinates = row['longitude'] + row['latitude']
+            if coordinates not in multi_lingual_addresses:
+                multi_lingual_addresses[coordinates] = {
+                    'street': {
+                        'municipality': turku,
+                    },
+                    'point': {
+                        'x': float(row['longitude']),
+                        'y': float(row['latitude']),
+                    },
+                    'address': {
+                        'number': row['street_number'],
+                    }
+                }
+
+            if row['city'].lower() == 'turku':
+                multi_lingual_addresses[coordinates]['street']['name'] = row['street']
+                multi_lingual_addresses[coordinates]['street']['name_en'] = row['street']
+            elif row['city'].lower() == 'åbo':
+                # As we expect Swedish names to come after Finnish ones, we always default
+                # to the Swedish name, but if there are a Swedish only street name, we
+                # fall back to Swedish for all name entries
+                if 'name' not in multi_lingual_addresses[coordinates]['street']:
+                    multi_lingual_addresses[coordinates]['street']['name'] = row['street']
+                    multi_lingual_addresses[coordinates]['street']['name_en'] = row['street']
+
+                multi_lingual_addresses[coordinates]['street']['name_sv'] = row['street']
+        return multi_lingual_addresses
+
+    def import_addresses(self):
+        file_path = os.path.join(self.data_path, 'turku_addresses.csv')
+        entries_created = 0
+
+        Street.objects.all().delete()
+        Address.objects.all().delete()
+
+        with open(file_path, encoding='latin-1') as csvfile:
+            address_reader = csv.DictReader(csvfile, delimiter=';', fieldnames=self.csv_field_names)
+            multi_lingual_addresses = self._create_address_mapping(address_reader)
+
+            for entry in multi_lingual_addresses.values():
+                self._import_address(entry)
+                entries_created += 1
+                if entries_created % 1000 == 0:
+                    self.logger.debug('row {} / ~24000'.format(entries_created))
```

Comment:
Could easily get the exact amount from `len(multi_lingual_addresses.values())` to handle possible changes in the future?

Reply:
True that, relic from before some refactoring, good point 👍